### PR TITLE
Wagtail 3.0 compatibility

### PR DESCRIPTION
**Pull request by nickmoreton**
### Add support for Wagtail 3.0

Support Ticket: https://projects.torchbox.com/projects/support-team/tickets/463

This also includes Django 4.0 support

It would be good if the example screenshots could be updated. I wonder if @kaedroho might have the original db or fixtures to do that?

Tasks

- [x] Update docs
- [x] Update Change Log

### CODE REVIEW

File: wagtail_ab_testing/wagtail_hooks.py [hunk 1]
```diff
@@ -52,15 +54,28 @@ class CreateAbTestActionMenuItem(ActionMenuItem):
     label = __("Save and create A/B Test")
     icon_name = 'people-arrows'
 
-    def is_shown(self, request, context):
-        if context['view'] != 'edit':
-            return False
+    if WAGTAIL_VERSION >= (2, 15):
+        # see if this is a pre-2.15
```
Comment (zerolab) [inline wagtail_ab_testing/wagtail_hooks.py:58]: nitpick: >= 2.15 is not pre-2.15 :)

File: wagtail_ab_testing/wagtail_hooks.py [hunk 2]
```diff
@@ -70,28 +85,54 @@ def register_create_abtest_action_menu_item():
...
-                    'can_create_abtest': request.user.has_perm('wagtail_ab_testing.add_abtest'),
-                }))
-            )
-
-        return ''
+    if WAGTAIL_VERSION >= (2, 15):
+        # see if this is a pre-2.15 
+        def render_html(self, context):
+            if 'page' in context:
+                return format_html(
```
Comment (zerolab) [inline wagtail_ab_testing/wagtail_hooks.py:92]: suggestion: move this to own method that accepts the user and context as params, then the two `render_html` definitions will be nice short, sweet and readable

File: testmanage.py [hunk 3]
```diff
@@ -8,6 +8,13 @@
 
 from django.core.management import execute_from_command_line
 
+from wagtail import VERSION as WAGTAIL_VERSION
+
+if WAGTAIL_VERSION >= (3, 0):
```
[the remark this answers is not in the document]
Comment (nickmoreton) [reply testmanage.py]: Thanks Brian, good point. All done.

File: CHANGELOG.md [hunk 4]
```diff
@@ -7,6 +7,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
```
Comment (William-Blackie) [inline CHANGELOG.md:9]: This is a major release, right? We need to add the new version number here.

File: testmanage.py [hunk 5]
```diff
@@ -8,6 +8,11 @@
 
 from django.core.management import execute_from_command_line
 
+try:
+    from wagtail.test.settings import STATIC_ROOT, MEDIA_ROOT
+except ImportError:
```
Comment (William-Blackie) [inline testmanage.py:13]: Issue: Here we're using the try except to implicitly work as a version check. Why not just check the version properly? It's not clear that this check is for 3.0 and in the future could break functionality. 
```
from wagtail import VERSION as WAGTAIL_VERSION

if WAGTAIL_VERSION >= (3, 0):
    from wagtail.tests.settings import STATIC_ROOT, MEDIA_ROOT
else:
     from wagtail.tests.settings import STATIC_ROOT, MEDIA_ROOT
```

File: wagtail_ab_testing/api.py [hunk 6]
```diff
@@ -2,7 +2,11 @@
 from rest_framework import fields, routers, serializers, status, viewsets
 from rest_framework.decorators import action
 from rest_framework.response import Response
-from wagtail.core.models import Page, Site
+
+try:
+    from wagtail.models import Page, Site
+except ImportError:
+    from wagtail.core.models import Page, Site
```
Comment (William-Blackie) [inline wagtail_ab_testing/api.py:9]: Issue: Implict checks.

File: wagtail_ab_testing/events.py [hunk 7]
```diff
@@ -1,6 +1,9 @@
 from django.utils.translation import gettext_lazy as __
 
-from wagtail.core import hooks
+try:
+    from wagtail import hooks
+except ImportError:
+    from wagtail.core import hooks
```
Comment (William-Blackie) [inline wagtail_ab_testing/events.py:6]: Issue: Implicit checks.

File: wagtail_ab_testing/views.py [hunk 8]
```diff
@@ -226,14 +237,28 @@ def __init__(self, request, **kwargs):
...
     def render_html(self):
-        return render_to_string(self.template, {
-            'default_menu_item': self.default_item.render_html(self.request, self.context),
-            'show_menu': bool(self.menu_items),
-            'rendered_menu_items': [
-                menu_item.render_html(self.request, self.context)
-                for menu_item in self.menu_items
-            ],
-        }, request=self.request)
+        if WAGTAIL_VERSION >= (3, 0):
```
Comment (William-Blackie) [inline wagtail_ab_testing/views.py:240]: Thought: Why not just pin this to >= 2.15 then?

**Review by William-Blackie — CHANGES_REQUESTED** [review]
Hi @nickmoreton, could we change these try-except blocks to use properly checked version imports?

File: testmanage.py (hunk 5 above)
Comment (nickmoreton) [reply testmanage.py]: Fixed

File: wagtail_ab_testing/api.py (hunk 6 above)
Comment (nickmoreton) [reply wagtail_ab_testing/api.py:11]: Fixed

File: wagtail_ab_testing/events.py (hunk 7 above)
Comment (nickmoreton) [reply wagtail_ab_testing/events.py:6]: Fixed

File: wagtail_ab_testing/views.py (hunk 8 above)
Comment (nickmoreton) [reply wagtail_ab_testing/views.py]: Changed

File: CHANGELOG.md (hunk 4 above)
Comment (nickmoreton) [reply CHANGELOG.md:9]: I think the final release number should now be 1.0.0

**Review by bcdickinson — CHANGES_REQUESTED** [review]
This all looks good to me with the exception of a rogue `pytz` import.

File: wagtail_ab_testing/test/tests/test_compare_draft.py [hunk 9]
```diff
@@ -1,7 +1,15 @@
 from django.test import TestCase
 from django.urls import reverse
-from wagtail.core.models import Page
-from wagtail.tests.utils import WagtailTestUtils
+from pytz import VERSION
```
[the remark this answers is not in the document]
Comment (nickmoreton) [reply wagtail_ab_testing/test/tests/test_compare_draft.py]: Ah, no it's not used is it. I have removed it. Thanks @bcdickinson

**Comment by nickmoreton** [comment]
Hey @bcdickinson Let me know please if you are happy for me to merge this now and I can then move on with the [v4](https://github.com/torchbox/wagtail-ab-testing/pull/49) one

**Comment by nickmoreton** [comment]
Oh actually I don't have the correct access to merge it.

**Comment by Stormheg** [comment]
superseded by #52

**Pull request closed without merging**